ambuda/views/dictionaries.py

```diff
@@ -32,41 +44,58 @@ def _fetch_entries(version: str, query: str) -> list[str]:
         "vacaspatyam": xml.transform_vacaspatyam,
         "shabdakalpadruma": xml.transform_mw,
     }
-    fn = transforms.get(version, xml.transform_mw)
-    return [fn(r.value) for r in rows]
+
+    results = {}
+    for source_slug, source_entries in entries.items():
+        fn = transforms.get(source_slug, xml.transform_mw)
+        html_blobs = [fn(e.value) for e in source_entries]
+        results[source_slug] = html_blobs
+    return results
 
 
 @bp.route("/")
 def index():
     """Show the dictionary lookup tool."""
-    return render_template("dictionaries/index.html")
+    dictionaries = _get_dictionary_data()
+    return render_template("dictionaries/index.html", dictionaries=dictionaries)
 
 
-@bp.route("/<slug>/")
-def version(slug):
-    dictionaries = q.dictionaries()
-    if slug not in dictionaries:
+@bp.route("/<list:slugs>/")
+def sources(slugs):
+    if not all(s in _get_dictionary_data() for s in slugs):
         abort(404)
-    # TODO: set chosen dictionary as UX view
+    # TODO: set chosen dictionaries as UX view
     return redirect(url_for("dictionaries.index"))
 
 
-@bp.route("/<version>/<query>")
-def entry(version, query):
+@bp.route("/<list:sources>/<query>")
+def entry(sources, query):
     """Show a specific dictionary entry."""
-    dictionaries = q.dictionaries()
-    if version not in dictionaries:
+    dictionaries = _get_dictionary_data()
+    if not sources or not all(s in dictionaries for s in sources):
+        # Abort if sources aren't reasonable.
         abort(404)
 
-    entries = _fetch_entries(version, query)
-    return render_template("dictionaries/index.html", query=query, entries=entries)
+    entries = _fetch_entries(sources, query)
+    return render_template(
+        "dictionaries/index.html",
+        query=query,
+        entries=entries,
+        dictionaries=dictionaries,
+    )
 
 
-@api.route("/dictionaries/<version>/<query>")
-def entry_htmx(version, query):
-    dictionaries = q.dictionaries()
-    if version not in dictionaries:
+@api.route("/dictionaries/<list:sources>/<query>")
+def entry_htmx(sources, query):
+    dictionaries = _get_dictionary_data()
+    if not sources or not all(s in _get_dictionary_data() for s in sources):
+        # Abort if sources aren't reasonable.
```

Comment:
I know this gets populated from our own frontend code so errors are unlikely, but just to be defensive / fail less completely: maybe allow at least one to be ok? I mean something like:

```
dictionaries = _get_dictionary_data()
sources = [s for s in sources if s in _get_dictionary_data()]
if not sources:
    abort(404)
```

Reply:
ah, good point! done.